src/client/Microsoft.Identity.Client/Internal/Requests/ClientCredentialRequest.cs

```diff
@@ -189,16 +204,51 @@ private async Task<AuthenticationResult> SendTokenRequestToAppTokenProviderAsync
 
         private async Task<MsalAccessTokenCacheItem> GetCachedAccessTokenAsync()
         {
-            MsalAccessTokenCacheItem cachedAccessTokenItem = await CacheManager.FindAccessTokenAsync().ConfigureAwait(false);
+            // Fetch the cache item (could be null if none found).
+            MsalAccessTokenCacheItem cacheItem =
+                await CacheManager.FindAccessTokenAsync().ConfigureAwait(false);
 
-            if (cachedAccessTokenItem != null && !_clientParameters.ForceRefresh)
+            // If the item fails any checks (null, or hash mismatch),
+            if (!ShouldUseCachedToken(cacheItem))
             {
-                AuthenticationRequestParameters.RequestContext.ApiEvent.IsAccessTokenCacheHit = true;
-                Metrics.IncrementTotalAccessTokensFromCache();
-                return cachedAccessTokenItem;
+                return null;
```

Comment:
Should we add a warning to let the user know that the hash they provided was not matched to an existing hash?

Reply:
yes there is one at `AcquireTokenForClientParameters`

Reply:
Actually, we are [accepting null or empty string as a valid input](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/5179#discussion_r1989085160). @gladjohn 

Besides, an input hash not matching any existing token is considered normal (especially in Service Fabric's scenario); that is what we designed this WithAccessTokenSha256ToRefresh() for: to silently serve a cached token to a request with mismatching hash. @Robbie-Microsoft